src/snowflake/snowpark/_internal/ast.py

```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
+#
+
+import base64
+import itertools
+import sys
+import uuid
+from dataclasses import dataclass
+from typing import Callable, Tuple
+
+import snowflake.snowpark._internal.proto.generated.ast_pb2 as proto
+from snowflake.snowpark.version import VERSION
+
+# The current AST version number (generated by the DSL).
+CLIENT_AST_VERSION = proto.__Version__.MAX_VERSION
+
+
+@dataclass
+class TrackedCallable:
+    var_id: int
+    func: Callable
+
+
+class AstBatch:
+    """
+    A batch of AST statements. This class is used to generate AST requests.
+
+    The core statement types are:
+    - Assign: Creates a new variable and assigns a value to it.
+    - Eval: Evaluates a variable.
+    """
+
+    # Function used to generate request IDs. This is overridden in some tests.
+    generate_request_id = uuid.uuid4
+
+    def __init__(self, session) -> None:
+        """
+        Initializes a new AST batch.
+
+        Args:
+            session: The Snowpark session.
+        """
+        self._session = session
+        self.reset_id_gen()
+        self._init_batch()
+
+        # Track callables in this dict (memory id -> TrackedCallable).
+        self._callables = {}
+
+    def reset_id_gen(self):
+        """Resets the ID generator."""
+        self._id_gen = itertools.count(start=1)
+
+    def assign(self, symbol=None):
+        """
+        Creates a new assignment statement.
+
+        Args:
+            symbol: An optional symbol to name the new variable.
+        """
+        stmt = self._request.body.add()
+        # TODO: extended BindingId spec from the branch snowpark-ir.
+        stmt.assign.uid = next(self._id_gen)
+        stmt.assign.var_id.bitfield1 = stmt.assign.uid
+        stmt.assign.symbol.value = symbol if isinstance(symbol, str) else ""
+        return stmt.assign
+
+    def eval(self, target):
+        """
+        Creates a new evaluation statement.
+
+        Args:
+            target: The variable to evaluate.
+        """
+        stmt = self._request.body.add()
+        stmt.eval.uid = next(self._id_gen)
+        stmt.eval.var_id.CopyFrom(target.var_id)
+
+    def flush(self) -> Tuple[str, str]:
+        """Ties off a batch and starts a new one. Returns the tied-off batch."""
+        batch = str(base64.b64encode(self._request.SerializeToString()), "utf-8")
+        self._init_batch()
+        return (str(self._request_id), batch)
+
+    def _init_batch(self):
+        self._request_id = AstBatch.generate_request_id()  # Generate a new unique ID.
```

Comment:
add a single line comment to explain what init do here.

Reply:
Done.